java/arcs/core/analysis/InformationFlow.kt

```diff
@@ -235,6 +246,166 @@ class InformationFlow private constructor(
         )
     }
 
+    /** Semantics of a [TypeVariables] in a particle. */
+    data class TypeVariableSemantics(
+        val claims: List<Claim>,
+        val inaccessiblePaths: List<AccessPath>
+    )
+
+    /** Information about restrictions on [AccessPath] induced by a [TypeVariable]. */
+    data class AccessPathRestrictions(
+        val typeVariable: TypeVariable,
+        val prefix: AccessPath,
+        val accessibleSelectors: Set<List<AccessPath.Selector>>,
+        val inaccessibleSelectors: Set<List<AccessPath.Selector>>
+    )
+
+    /** Returns the restrictions induced by [TypeVariables] in the type. */
+    private fun getAccessPathRestrictions(
+        specType: Type,
+        resolvedType: Type,
+        prefix: AccessPath
+    ): List<AccessPathRestrictions> {
+        if (specType.tag == Tag.TypeVariable) {
+            specType as TypeVariable
+            val resolvedTypeSelectors = resolvedType.accessPathSelectors(partial = true)
+            val accessibleSelectors =
+                specType.constraint?.accessPathSelectors(partial = true) ?: resolvedTypeSelectors
+            val inaccessibleSelectors = resolvedTypeSelectors.minus(accessibleSelectors)
+            return listOf(
+                AccessPathRestrictions(
+                    specType,
+                    prefix,
+                    accessibleSelectors,
+                    inaccessibleSelectors
+                )
+            )
+        }
+        require(specType.tag == resolvedType.tag) {
+            "Type of a connection in particle spec is incompatible with " +
+            "the type of the corresponding connection in a particle."
+        }
+        return when (specType.tag) {
+            Tag.Collection -> getAccessPathRestrictions(
+                (specType as CollectionType<*>).collectionType,
+                (resolvedType as CollectionType<*>).collectionType,
+                prefix
+            )
```

Comment:
It's kind of annoying how many times this exact thing is repeated `getAccessPathRestrictions(specType.foo, resolvedType.foo, prefix)` but we're unable to de-dupe the logic... 

I wonder if it's worth lifting `.containedType` out into its own interface (e.g. `ContainerType`), so that we have a way of talking about all of them in the same way?

Mainly food for thought... Would be nice to clean up some of these when-expressions though.

Reply:
Yeah, I agree. It is certainly annoying. `.containedType` is good idea, but that will require some non-trivial refactoring.